Framework/API/src/ExperimentInfo.cpp

```diff
@@ -1018,39 +1085,54 @@ DetectorInfo &ExperimentInfo::mutableDetectorInfo() {
+/// Return a reference to the Beamline::SpectrumInfo object. Helper for
+/// API::SpectrumInfo, do not use this.
+const Beamline::SpectrumInfo &ExperimentInfo::internalSpectrumInfo() const {
   populateIfNotLoaded();
-  std::call_once(m_defaultDetectorGroupingCached,
-                 &ExperimentInfo::cacheDefaultDetectorGrouping, this);
-
-  return m_detgroups.size();
+  if (!m_spectrumInfo) {
+    std::lock_guard<std::mutex> lock{m_spectrumInfoMutex};
+    if (!m_spectrumInfo)
+      cacheDefaultDetectorGrouping();
+  }
+  return *m_spectrumInfo;
 }
 
-/** Returns a set of detector IDs for a group.
+/** Return a reference to the SpectrumInfo object.
  *
- * This is a virtual method. The default implementation returns grouping
- * information cached in the ExperimentInfo. This is used in MDAlgorithms. The
- * more common case is handled by the overload of this method in
- * MatrixWorkspace. The purpose of this method is to be able to construct
- * SpectrumInfo based on an ExperimentInfo object, including grouping
- * information. Grouping information can be cached in ExperimentInfo, or can be
- * obtained from child classes (MatrixWorkspace). */
-const std::set<detid_t> &
-ExperimentInfo::detectorIDsInGroup(const size_t index) const {
+ * Any modifications of the instrument or instrument parameters will invalidate
+ * this reference.
+ */
+const SpectrumInfo &ExperimentInfo::spectrumInfo() const {
   populateIfNotLoaded();
-  std::call_once(m_defaultDetectorGroupingCached,
-                 &ExperimentInfo::cacheDefaultDetectorGrouping, this);
+  if (!m_spectrumInfoWrapper) {
+    std::lock_guard<std::mutex> lock{m_spectrumInfoMutex};
+    if (!m_spectrumInfoWrapper)
+      m_spectrumInfoWrapper = Kernel::make_unique<SpectrumInfo>(*this);
+  }
+  return *m_spectrumInfoWrapper;
+}
 
-  return m_detgroups.at(index);
+/** Return a non-const reference to the SpectrumInfo object. Not thread safe.
+ */
+SpectrumInfo &ExperimentInfo::mutableSpectrumInfo() {
+  populateIfNotLoaded();
+  // Creating SpectrumInfo with a non-const reference to a MatrixWorkspace will
+  // call ExperimentInfo::mutableDetectorInfo() which will later be used by
+  // modifications. This will trigger a copy if required. Note that the
+  // following happens internally:
+  // 1. make_unique creates a new SpectrumInfo, which calls
+  // ExperimentInfo::mutableDetectorInfo(). In the latter method, the reference
+  // count to the ParameterMap is typically not 1, so
+  // invalidateInstrumentReferences() is called.
+  // 2. invalidateInstrumentReferences() resets m_spectrumInfoWrapper, releasing
+  // any parameterized detectors and thus dropping any unneeded references to
+  // the ParameterMap.
+  // 3. Construction of SpectrumInfo continues and the result is assigned to
+  // m_spectrumInfoWrapper.
+
+  // No locking here since this non-const method is not thread safe.
+  m_spectrumInfoWrapper = Kernel::make_unique<SpectrumInfo>(*this);
```

Comment:
I think once we store positions in `Beamline::DetectorInfo` we would be able to remove the caching of detectors in `API::SpectrumInfo` and `API::DetectorInfo`. That would require changing `API::SpectrumInfo::detector` to return a `shared_ptr` instead of a reference (and some refactoring). Once that is done, the only remaining hurdle is the caching of source and sample positions. If we find a way to solve also that, we can remove the weird reinitialization mechanism.